pkg/pillar/cmd/zedrouter/flowstats.go

```diff
@@ -0,0 +1,706 @@
+					DstPort: int32(tuple.DstPort),
+					Proto:   int32(tuple.Proto),
+				}
+				flowrec := types.FlowRec{
+					Flow:      flowtuple,
+					Inbound:   !tuple.AppInitiate,
+					ACLID:     int32(aclattr.aclNum),
+					Action:    aclattr.action,
+					StartTime: tuple.TimeStart,
+					StopTime:  tuple.TimeStop,
+					TxBytes:   int64(tuple.SendBytes),
+					TxPkts:    int64(tuple.SendPkts),
+					RxBytes:   int64(tuple.RecvBytes),
+					RxPkts:    int64(tuple.RecvPkts),
+				}
+
+				flowdata.Flows = append(flowdata.Flows, flowrec)
+			}
+
+			if len(flowdata.Flows) == 0 {
+				continue
+			}
+			// temp print out DNS to Address Req/Reply
+			// only per bridge snoop for dns
+			var dnsrec [2]map[string]dnsEntry
+			dnsrec[0] = make(map[string]dnsEntry) // store IPv4 addresses from dns
+			dnsrec[1] = make(map[string]dnsEntry) // store IPv6 addresses from dns
+			bnNum, err := bridgeStrToNum(bnx)
+			if err != nil {
+				continue
+			}
+
+			// get the bringe-X dns request/replies
+			dnssys[bnNum].Lock()
+			defer dnssys[bnNum].Unlock()
+			for _, dnsdata := range dnssys[bnNum].Snoop {
+				// unique by domain name, latest reply overwrite previous ones
+				if dnsdata.isIPv4 {
+					dnsrec[0][dnsdata.DomainName] = dnsdata
+				} else {
+					dnsrec[1][dnsdata.DomainName] = dnsdata
+				}
+			}
+			for idx := range dnsrec {
+				for _, dnsRec := range dnsrec[idx] {
+					// temp print out all unique dns replies for the bridge
+					log.Infof("!!FlowStats: DNS time %v, domain %s, appIP %v, count %d, Answers %v",
+						dnsRec.TimeStamp, dnsRec.DomainName, dnsRec.AppIP, dnsRec.ANCount, dnsRec.Answers)
+
+					dnsrec := types.DNSReq{
+						HostName:    dnsRec.DomainName,
+						Addrs:       dnsRec.Answers,
+						RequestTime: dnsRec.TimeStamp.UnixNano(),
+					}
+					dnsPacked = true
+					flowdata.DNSReqs = append(flowdata.DNSReqs, dnsrec)
+				}
+			}
+			// flow record done for the bridge/app
+			//flowdata.CastFlowStatus
+			flowKey := scope.UUID.String() + scope.NetUUID.String()
+			pub.Publish(flowKey, &flowdata)
+		}
+	}
+
+	// remove the dns data already uploaded
+	if dnsPacked {
+		for bnx := range instData.bnNet {
+			bnNum, err := bridgeStrToNum(bnx)
+			if err != nil {
+				continue
+			}
+			dnssys[bnNum].Snoop = nil
+			log.Infof("!!FlowStats: clear dns record for bn%d", bnNum)
+		}
+	}
+}
+
+// conntrack flow of two uni-directional stats into one
+// bireditional flow stats
+func flowMergeProcess(entry *netlink.ConntrackFlow, instData networkAttrs) flowStats {
+	var ipFlow flowStats
+	var forwSrcApp, forwDstApp, backSrcApp, backDstApp bool
+	var AppNum int
+
+	// for the current phase of implementation, we ignore the flowstats which has not timed out yet
+	ipflowTimeOut := int32(entry.TimeOut)
+	if ipflowTimeOut > timeoutSec {
+		return ipFlow
+	}
+
+	ipFlow.aclNum = entry.Mark & markMask
+	ipFlow.appNum = uint8(entry.Mark >> appShiftBits)
+	AppNum = int(ipFlow.appNum)
+
+	//AppNum := 1 // XXX hack, = ipFlow.appNum later
```

Comment:
Looks like these two commented lines are not required anymore. Might make sense to just delete them?

Reply:
will do